contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -3851,16 +3871,215 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+							foreach(lc, grant->grantees)
+							{
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								/* Don't add/update an entry, if the permission is granted on column list.*/
+								if (ap->cols == NULL)
+									add_or_update_object_in_bbf_schema(logical_schema, obj, privilege, rol_spec->rolename, OBJ_RELATION, true, NULL);
+							}
+						}
+						else
+						{
+							foreach(lc, grant->grantees)
+							{
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								/*
+								 * If permission on schema exists, don't revoke any permission from the object.
+								 */
+								if (!privilege_exists_in_bbf_schema_permissions(logical_schema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, rol_spec->rolename, false))
+									exec_pg_command = true;
+
+								/* Don't update an entry, if the permission is granted on column list.*/
+								if (ap->cols == NULL)
+									update_privileges_of_object(logical_schema, obj, privilege, rol_spec->rolename, OBJ_RELATION, false);
+							}
+						}
+					}
+					if (exec_pg_command)
+						call_prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv, dest, qc);
+					return;
+				}
+				else if ((grant->objtype == OBJECT_PROCEDURE) || (grant->objtype == OBJECT_FUNCTION))
+				{
+					ObjectWithArgs  *ob = (ObjectWithArgs *) linitial(grant->objects);
+					ListCell   *lc;
+					ListCell	*lc1;
+					bool exec_pg_command = false;
+					const char *logicalschema = NULL;
+					char *funcname = NULL;
+					const char *obj_type = NULL;
+					Oid func_oid = LookupFuncWithArgs(OBJECT_ROUTINE, ob, true);
+					const char *func_args = NULL;
+					if (OidIsValid(func_oid))
+						func_args = gen_func_arg_list(func_oid);
+					if (grant->objtype == OBJECT_FUNCTION)
+						obj_type = OBJ_FUNCTION;
+					else
+						obj_type = OBJ_PROCEDURE;
+					if (list_length(ob->objname) == 1)
+					{
+						Node *func = (Node *) linitial(ob->objname);
+						funcname = strVal(func);
+						logicalschema = get_authid_user_ext_schema_name(dbname, current_user);
+					}
+					else
+					{
+						Node *schema = (Node *) linitial(ob->objname);
+						char *schemaname = strVal(schema);
+						Node *func = (Node *) lsecond(ob->objname);
+						logicalschema = get_logical_schema_name(schemaname, true);
+						funcname = strVal(func);
+					}
+
+					/* If ALL PRIVILEGES is granted/revoked. */
+					if (list_length(grant->privileges) == 0)
+					{
+						/*
+						 * Case: When ALL PRIVILEGES is revoked internally during create function.
+						 * pstmt->stmt_len = 0 means it is an implicit REVOKE statement issued at the time of create function/procedure.
+						 * For more details, please refer revoke_func_permission_from_public().
+						 * If schema entry exists in the catalog, implicitly grant permission on the new object to the user.
+						 */
+						if ((pstmt->stmt_len == 0) && privilege_exists_in_bbf_schema_permissions(logicalschema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, NULL, true))
```

Comment:
Can there be a case when (pstmt->stmt_len == 0) but privilege_exists_in_bbf_schema_permissions(logicalschema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, NULL, true) = false?

Reply:
Yes, it happens for a case when we create a function in a schema where there is no privilege granted by user.